Zend/Optimizer/zend_optimizer.c

```diff
@@ -831,6 +831,49 @@ zend_class_entry *zend_optimizer_get_class_entry_from_op1(
 	return NULL;
 }
 
+const zend_class_constant *zend_fetch_class_const_info(
+	const zend_script *script, const zend_op_array *op_array, const zend_op *opline, bool *is_prototype) {
+	const zend_class_entry *ce = NULL;
+	bool is_static_reference = false;
+
+	if (!opline || !op_array || !script || opline->op2_type != IS_CONST || Z_TYPE(ZEND_OP2_LITERAL(opline)) != IS_STRING) {
+		return NULL;
+	}
+	if (opline->op1_type == IS_CONST) {
+		zval *op1 = CRT_CONSTANT(opline->op1);
+		if (Z_TYPE_P(op1) == IS_STRING) {
+			ce = zend_optimizer_get_class_entry(script, op_array, Z_STR_P(op1 + 1));
+		}
+	} else if (opline->op1_type == IS_UNUSED && op_array->scope && !(op_array->scope->ce_flags & ZEND_ACC_TRAIT)) {
+		int fetch_type = (int) opline->op1.num & ZEND_FETCH_CLASS_MASK;
+		if (fetch_type == ZEND_FETCH_CLASS_SELF) {
+			ce = op_array->scope;
+		} else if (fetch_type == ZEND_FETCH_CLASS_STATIC) {
+			ce = op_array->scope;
+			is_static_reference = true;
+		} else if (fetch_type == ZEND_FETCH_CLASS_PARENT) {
+			if (op_array->scope->ce_flags & ZEND_ACC_LINKED) {
+				ce = op_array->scope->parent;
+			}
+		}
+	}
+	if (!ce || (ce->ce_flags & ZEND_ACC_TRAIT)) {
```

Comment:
I believe we cannot reach `(ce->ce_flags & ZEND_ACC_TRAIT)`. It's already guarded above through `!(op_array->scope->ce_flags & ZEND_ACC_TRAIT)`, and parents can never be traits.

Reply:
Removing this condition makes the `Zend/tests/traits/constant_004.phpt` test fail, so I'll skip this

Reply:
Ah, right. It's because of the `opline->op1_type == IS_CONST` branch where traits aren't handled yet. This is fine then.